README.md
powdR: Full Pattern Summation of X-Ray Powder Diffraction Data
================

<!-- README.md is generated from README.Rmd. Please edit that file -->

[![Build
Status](https://travis-ci.org/benmbutler/powdR.svg?branch=master)](https://travis-ci.org/benmbutler/powdR)
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/powdR)](https://CRAN.R-project.org/package=powdR)
[![Total\_Downloads](https://cranlogs.r-pkg.org/badges/grand-total/powdR)](https://cran.r-project.org/package=powdR)

## Overview

`powdR` is an implementation of the full pattern summation approach to
quantitative mineralogy from X-ray powder diffraction data (Chipera and
Bish 2002, 2013; Eberl 2003).

`powdR` has several advantages over the excel based implementations of
full pattern summation such as FULLPAT (Chipera and Bish 2002) and
RockJock (Eberl 2003). First, computation is faster and, when
quantifying multiple samples, can be used in combination with other
packages (e.g [`foreach`](https://cran.r-project.org/package=foreach)
and [`doParallel`](https://cran.r-project.org/package=doParallel)) for
parallel processing. Secondly, powdR can be run via a Shiny web
application, which offers a user friendly interface for fast and
iterative mineral quantification. Lastly, R represents a powerful tool
for data manipulation, allowing users to creatively adapt, pre-treat and
visualise their data.

## Installation

The stable version of `powdR` is on CRAN:

``` r
install.packages("powdR")
```

Alternatively, the development version can be downloaded from GitHub

``` r
#Install devtools if you don't already have it on your machine
install.packages("devtools")

devtools::install_github('benmbutler/powdR')
```

## Usage

``` r
library(powdR)
#> powdR: Full Pattern Summation of X-Ray Powder Diffraction Data

#Load some soils to quantify
data(soils)

#Load a powdRlib reference library of pure patterns
data(minerals)

#Quantify a sample
q <-  fps(lib = minerals,
          smpl = soils$sandstone,
          refs = minerals$phases$phase_id,
          std = "QUA.1")
#> 
#> -Aligning sample to the internal standard
#> -Interpolating library to same 2theta scale as aligned sample
#> -Optimising...
#> -Removing negative coefficients and reoptimising...
#> -Computing phase concentrations
#> -Internal standard concentration unknown. Assuming phases sum to 100 %
#> ***Full pattern summation complete***

#Inspect the phase concentrations (summarised by name)
q$phases_grouped
#>       phase_name phase_percent
#> 1         Illite        1.3290
#> 2     K-feldspar        1.3020
#> 3      Kaolinite        1.3697
#> 4 Organic-Matter       41.1591
#> 5    Plagioclase        1.1632
#> 6         Quartz       53.6769

#Inspect the quantification
plot(q, wavelength = "Cu")
```

![](man/figures/README-example-1.png)<!-- -->

Alternatively, `plot(q, wavelength = "Cu", interactive = TRUE)` provides
an interactive plot for better inspection of the fit. More detailed
usage instructions are provided in the package vignette.

## The powdR Shiny app

To run `powdR` via the Shiny app, use `run_powdR()`. This loads the
application in your default web browser. The application has six tabs:

1.  **Reference Library Builder:** Allows you to create and export a
    `powdRlib` reference library from two .csv files: one for the XRPD
    measurements, and the other for the ID, name and reference intensity
    ratio of each pattern.
2.  **Reference Library Viewer:** Facilitates quick inspection of the
    phases within a `powdRlib` reference library.
3.  **Reference Library Editor:** Allows the user to easily subset a
    `powdRlib` reference library .
4.  **Full Pattern Summation:** A user friendly interface for manual and
    automated full pattern summation of single samples.
5.  **Results Viewer/Editor:** Allows for results from previously saved
    `powdRfps` and `powdRafps` objects to be viewed and edited via
    addition or removal of reference patterns.
6.  **Help** Provides a series of video tutorials (via YouTube)
    detailing the use of the `powdR` Shiny application.

**Please note that the `powdR` Shiny app may not work in the Microsoft
Edge internet browser.**

## References

<div id="refs" class="references">

<div id="ref-Chipera2002">

Chipera, Steve J., and David L. Bish. 2002. “FULLPAT: A full-pattern
quantitative analysis program for X-ray powder diffraction using
measured and calculated patterns.” *Journal of Applied Crystallography*
35 (6): 744–49. <https://doi.org/10.1107/S0021889802017405>.

</div>

<div id="ref-Chipera2013">

———. 2013. “Fitting Full X-Ray Diffraction Patterns for Quantitative
Analysis: A Method for Readily Quantifying Crystalline and Disordered
Phases.” *Advances in Materials Physics and Chemistry* 03 (01): 47–53.
<https://doi.org/10.4236/ampc.2013.31A007>.

</div>

<div id="ref-Eberl2003">

Eberl, D. D. 2003. “User’s guide to ROCKJOCK - A program for determining
quantitative mineralogy from powder X-ray diffraction data.” Boulder,
CA: USGS.

</div>

</div>


NEWS.md
# powdR 1.2.5

## New features
# `fps()` and `afps()` now accept diffraction data that has negative values for count
  intensities. In such cases Rwp cannot be used as the objective function and R will
  be used as the default instead.
# The `rwp` item in the outputs from `fps()` and `afps()` has been renamed `obj`, which
  contains a named vector of the values for three objective parameters: Rwp, R and
  Delta.
# `summarise_mineralogy` now accepts two additional arguments: `r` and `delta` which are
  logical parameters used to specify whether the R and Delta objective parameters,
  respectively,are included in the summary table.

# powdR 1.2.4

## New features
# `powdRlib()` now accepts a logical `check_names` argument. If `TRUE` (the default) then
  the names of the variables are checked to ensure that they are syntactically valid and
  are not duplicated.

## Bug fixes
# Default value for the normalise argument in `fps()` and `afps()` is now supplied
  (`FALSE`)
# `powdRlib()` now ensures that the `phases` object is a dataframe.

# powdR 1.2.3

## New features
* New function `regroup()` allows for an alternative mineral grouping structure to be applied to
  `powdRfps` and `powdRafps` objects.
* New function `tth_transform()` allows 2theta transformation between different monochromatic X-ray
  wavelengths.
* `fps()` and `afps()` now accept a logical `normalise` argument, which allows the internal standard
  concentration to be omitted and phase concentrations subsequently normalised to sum to 100 %.

# powdR 1.2.2

## New features
* `fps()` now accepts the `force` argument, forcing phases to remain in the final output even if their
  coefficients are negative.
* `fps()` and `afps()` will now stop if any of the phases specified in the `refs` argument are not
  in the library.
* Warning messages now generally do not include the call.
* Progress bar now included in the Shiny app when computing `fps()` and `afps()`.
* `summarise_mineralogy()` now accepts single samples (i.e. a list of 1 `powdRfps` or `powdRafps` object).

## Bug fixes
* `utils` no longer in imports (hence fixing associated note in CRAN checks).

# powdR 1.2.1

## New features
* `plot()` methods for `powdRfps` and `powdRafps` objects now include `mode`
  and `xlim` arguments, allowing for different plot types and x-axis adjustment.

## Bug fixes
* Dependency on packages `DT` and `shinyWidgets` now defined in namespace.
* Bug stopping the shiny app working in shiny versions >1.3.2 now fixed.
* Couple of other small edits to phase selection in the shiny app that should
  keep things simple.
* `powdRlib()` no longer orders the reference patterns alphabetically, and instead
  retains the original order that they are supplied in.
* Negative values from natural spline interpolation now removed from data as these
  can prevent the optimisation routines from working.
* Phases are no longer ordered alphabetically in outputs from `fps()` and `afps()`.

# powdR 1.2.0

## New features
* `fps()` and `afps()` no longer require the `shift_res` argument.
* Shifting is now calculated by optimisation of the objective function.
* Natural splines are now used instead of linear splines throughout `fps()` and `afps()`.
* The Full Pattern Summation tab in the Shiny application has been updated, as has the video
  tutorial for it.

# powdR 1.1.0

## New features
* The `refs` argument of `fps()` and `subset()` now accepts phase names as well as phase ID's. For example, if
  the phase name "Quartz" in supplied, then all phase ID's associated with Quartz will be selected.
  
* Similarly, the `force` argument of `afps()` now accepts both phase names and phase ID's.

* `summarise_mineralogy()` now contains an optional `rwp` argument (default = `FALSE`). This is a logical
  parameter used to define whether the Rwp should be included in the summary table as a measure of the
  difference between the measured and fitted patterns.
  
* When the `std_conc` argument is supplied to `fps()` or `afps()`, the computed phase concentrations now
  include that of the internal standard.

# powdR 1.0.0

## New features
*  Outputs from `fps()` and `afps()` (`powdRfps` and `powdRafps` objects, respectively) contain
   an `inputs` component. This provides a list of each of the arguments (including defaults) used to
   produce the fit.

*  `summarise_mineralogy()` is a new function that creates a summary table from lists containing
   multiple `powdRfps` and/or `powdRafps` objects.

*  A comprehensive reference library of pure phases from the RockJock computer software is now
   provided as an example `powdRlib` object called `rockjock`. This library covers most clay,
   non-clay and amorphous phases that may be encountered in soil samples. The library can be
   loaded into the global environment via `data(rockjock)`. Data of synthetic mineral mixtures
   are also now provided in the `rockjock_mixtures` data, which can be used to test the accuracy
   of full pattern summation via the `fps()` and `afps()` functions. 

*  `fps()` and `afps()` now accept "L-BFGS-B" in the `solver` argument. If selected, this uses
   L-BFGS-B optimisation constrained so that parameters cannot be lower than zero.

*  `fps()` now contains an optional `shift` argument, identical to that already implemented in
   `afps()`. This defines the 2$\theta$ range within with a grid-search algorithm can optimise the
    alignment of standards to the sample. If not defined in the function call it defaults to 0.
    
*  `fps()` and `afps()` now have a `shift_res` argument which accepts a single integer to define
   the increase in resolution used during grid search shifting. Higher values facilitate finer
   shifts at the expense of longer computation. If not defined in the function call it defaults
   to 4.
   

R/powdR.R
#' powdR: Full Pattern Summation of X-Ray Powder Diffraction Data
#'
#' An implementation of the full pattern summation approach to quantitative mineralogy
#' from X-ray powder diffraction data (Chipera & Bish, 2002, 2013; Eberl, 2003).
#' @docType package
#' @name powdR
#' @author Benjamin Butler, The James Hutton Institute, Aberdeen, UK
#' @references
#' Chipera, S.J., Bish, D.L., 2013. Fitting Full X-Ray Diffraction Patterns for Quantitative Analysis:
#' A Method for Readily Quantifying Crystalline and Disordered Phases. Adv. Mater. Phys. Chem. 03, 47-53.
#' doi:10.4236/ampc.2013.31A007
#'
#' Chipera, S.J., Bish, D.L., 2002. FULLPAT: A full-pattern quantitative analysis program for X-ray powder
#' diffraction using measured and calculated patterns. J. Appl. Crystallogr. 35, 744-749.
#' doi:10.1107/S0021889802017405
#'
#' Eberl, D.D., 2003. User's guide to ROCKJOCK - A program for determining quantitative mineralogy from
#' powder X-ray diffraction data. Boulder, CA.
NULL
